pystiche_papers/ulyanov_et_al_2016/core.py

```diff
@@ -0,0 +1,205 @@
+from typing import Callable, Optional, Tuple, Union, cast
+
+import torch
+from torch import nn
+from torch.optim.lr_scheduler import ExponentialLR
+from torch.optim.optimizer import Optimizer
+from torch.utils.data import DataLoader
+
+import pystiche
+from pystiche.loss import PerceptualLoss
+from pystiche.optim import OptimLogger, default_transformer_epoch_optim_loop
+
+from ..utils import batch_up_image
+from .data import (
+    ulyanov_et_al_2016_content_transform,
+    ulyanov_et_al_2016_dataset,
+    ulyanov_et_al_2016_image_loader,
+    ulyanov_et_al_2016_images,
+    ulyanov_et_al_2016_style_transform,
+)
+from .loss import ulyanov_et_al_2016_perceptual_loss
+from .modules import UlyanovEtAl2016Transformer, ulyanov_et_al_2016_transformer
+from .utils import (
+    ulyanov_et_al_2016_lr_scheduler,
+    ulyanov_et_al_2016_optimizer,
+    ulyanov_et_al_2016_postprocessor,
+    ulyanov_et_al_2016_preprocessor,
+)
+
+__all__ = [
+    "ulyanov_et_al_2016_transformer",
+    "ulyanov_et_al_2016_perceptual_loss",
+    "ulyanov_et_al_2016_dataset",
+    "ulyanov_et_al_2016_image_loader",
+    "ulyanov_et_al_2016_training",
+    "ulyanov_et_al_2016_images",
+    "ulyanov_et_al_2016_stylization",
+    "ulyanov_et_al_2016_texture_synthesis",
+]
+
+
+def ulyanov_et_al_2016_training(
+    content_image_loader: DataLoader,
+    style: Union[str, torch.Tensor],
+    impl_params: bool = True,
+    instance_norm: bool = True,
+    stylization: bool = True,
+    transformer: Optional[UlyanovEtAl2016Transformer] = None,
+    criterion: Optional[PerceptualLoss] = None,
+    lr_scheduler: Optional[ExponentialLR] = None,
+    num_epochs: Optional[int] = None,
+    get_optimizer: Optional[Callable[[UlyanovEtAl2016Transformer], Optimizer]] = None,
+    quiet: bool = False,
+    logger: Optional[OptimLogger] = None,
+    log_fn: Optional[
+        Callable[[int, Union[torch.Tensor, pystiche.LossDict], float, float], None]
+    ] = None,
+) -> nn.Module:
+    if isinstance(style, str):
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```

Comment:
We can use `get_device` from `pystiche.misc` here 
```suggestion
        device = get_device()
```